buildkit: add support for exporting build output with `--output` or `-o` (#3823)
containers/buildah: docs/buildah-build.1.md

Set the OS of the image to be built, and that of the base image to be pulled, if the build uses one, instead of using the current operating system of the host.

**--output**, **-o**=""

Output destination (format: type=local,dest=path)

> By default, a local container image is created from the build result. The --output (or -o) flag allows you to override this behavior, and a specify a custom exporter. For example, custom exporters allow you to export the build artifacts as files on the local filesystem instead of a Container image, which can be useful for generating local binaries, code generation etc.

Reviewer: ```suggestion
By default, a local container image is created from the build result. The --output (or -o) option allows you to override this behavior and specify a custom exporter. For example, custom exporters allow you to export the build artifacts as files on the local filesystem instead of a Container image, which can be useful for generating local binaries, code generation, etc.
```

Author: fixed

> The value for --output is a CSV-formatted string defining the exporter type and options. Currently, local and tar exporters are supported. The local exporter writes the resulting build files to a directory on the client side. The tar exporter is similar but writes the files as a single tarball (.tar).

Reviewer: ```suggestion
The value for --output is a CSV-formatted string defining the exporter type and options. 

Valid _type_ values are:
- **local**: write the resulting build files to a directory on the client-side.
- **tar**: write the resulting files as a single tarball (.tar).
```

Reviewer: An example or two would be helpful.  Looks like a good quick blog post once complete.

Reviewer: do you handle a hyphen `-` with the tar option to write the output to stdout?  You might want to add that to the man page if so.

Author: Yes we handle `-` , added to docs with example.

Set the OS of the image to be built, and that of the base image to be pulled, if the build uses one, instead of using the current operating system of the host.

**--output**, **-o**=""

Output destination (format: type=local,dest=path)

> By default, a local container image is created from the build result. The --output (or -o) option allows you to override this behavior and specify a custom exporter. For example, custom exporters allow you to export the build artifacts as files on the local filesystem instead of a Container image, which can be useful for generating local binaries, code generation, etc.

Reviewer 2: No need to capitalize container here.

> The value for --output is a CSV-formatted string defining the exporter type and options.

Reviewer 3: CSV is a specific spreadsheet format. This is not quite that.

The dest= option is only documented with an example… maybe go all the way and say that it is a comma-separated sequence of key=value pairs, and make a list/table of the supported keys.

Author: fixed.

Set the OS of the image to be built, and that of the base image to be pulled, if the build uses one, instead of using the current operating system of the host.

**--output**, **-o**=""

Output destination (format: type=local,dest=path)

> By default, a local container image is created from the build result. The --output (or -o) option allows you to override this behavior and specify a custom exporter. For example, custom exporters allow you to export the build artifacts as files on the local filesystem instead of a container image, which can be useful for generating local binaries, code generation, etc.

Reviewer 2: This suggests that when this option is used, we don't commit a container image, but it looks like we still do.

Author: Fixed explicitly mentioned.

Reviewer 2: This still reads to me as though using this option causes the builder to not commit an image, as committing the image is the default behavior that's being overridden when this option is specified.

Author: fixed it after confirming from @nalind hence marking resolved again.

The value for --output is a comma-separated sequence of key=value pairs, defining the output type and options.

Supported _keys_ are:
- **dest**: Destination path for exported output. Valid value is absolute or relative path.
- **type**: Defines the type of output to be used. Valid values is documented below.

Valid _type_ values are:
- **local**: write the resulting build files to a directory on the client-side.
- **tar**: write the resulting files as a single tarball (.tar).

> If no type is specified, the value defaults to local. Use a hyphen as non-comma format (`-`) to write the output tarball to standard output (`STDOUT`).

Reviewer 3: Maybe
```suggestion
If no type is specified, the value defaults to local. Use a hyphen as non-comma format (`-`) to write the output tarball to standard output.
```
There is no user-facing `STDOUT` identifier / name, to my knowledge.

> If no type is specified, the value defaults to local. Use a hyphen as non-comma format (`-`) to write the output tarball to standard output (`STDOUT`).

Reviewer 3: This
- doesn’t document `dest=-` AFAICS, although the implementation now supports that.
- Says “non-comma format”, which is a short-hand I started using for the purposes of this review, but that is unclear without extra content.

Maybe something vaguely like
> ```
> - **dest**: … absolute or relative path; or, as a special case, `-` means the standard output.
> …
> Alternatively, instead of a comma-separated sequence, the value of **--output** can be just a destination (in the `**dest** format) (e.g. `--output some-path`, `--output -`)
> ```

Reviewer 3: The “alternatively” part should also document the implied `type` value…

… wait, `--output foo` means `type=local`, but `--output -` means `type=tar`? That needs to be explicitly documented as well.

Set the OS of the image to be built, and that of the base image to be pulled, if the build uses one, instead of using the current operating system of the host.

**--output**, **-o**=""

Output destination (format: type=local,dest=path)

The --output (or -o) option extends the default behavior of building a container image by allowing users to export the contents of the image as files on the local filesystem, which can be useful for generating local binaries, code generation, etc.

The value for --output is a comma-separated sequence of key=value pairs, defining the output type and options.

Supported _keys_ are:
- **dest**: Destination path for exported output. Valid value is absolute or relative path, `-` means the standard output.
- **type**: Defines the type of output to be used. Valid values is documented below.

Valid _type_ values are:
- **local**: write the resulting build files to a directory on the client-side.
- **tar**: write the resulting files as a single tarball (.tar).

If no type is specified, the value defaults to **local**.

> Alternatively, instead of a comma-separated sequence, the value of **--output** can be just a destination (in the `**dest** format) (e.g. `--output some-path`, `--output -`) where `--output some-path` is equivalent to **type=local** and `--output -` is equivalent to **type=tar**.

Reviewer 3: One last nit: `s/is equivalent to/is treated as if/` or something like that. `some-path` is not “equivalent” to a type value.

Or commit to this approach, and write “`--output some-path` is equivalent to `type=local,dest=some-path`”.

buildah bud --all-platforms --manifest myimage /tmp/mysrc

### Building an image using (--output) custom build output

> buildah build -o out .

Reviewer 3: (Personal rant: Ugh, so the syntax of --output is a set of key-value pairs, _except when it’s not_? I hate ambiguities.)

Author: this is just to match the short-hand syntax which `docker buildx` supports.

Reviewer 2: The integration test should probably check that we don't unintentionally start rejecting this variation.

Reviewer 3: (Marked unresolved to add that test, or to explicitly decide that we don’t need one.)

Author: added a test.

buildah bud --all-platforms --manifest myimage /tmp/mysrc

### Building an image using (--output) custom build output

> buildah build -o out .

Author: Edited man page above
